docs/documentation/maintenance-coordinator.md

```diff
@@ -0,0 +1,91 @@
+# Coordinator Model
+
+## Table of Contents
+
+<!-- MarkdownTOC depth=0 autolink="true" bracket="round" autoanchor="true" style="ordered" indent="   " -->
+1. [Role of the Coordinator](#role-of-a-coordinator)
+1. [Change Management](#change-management)
+1. [Issue Management](#issue-management)
+1. [Scenarios](#scenarios)
+1. [What can be expected of a coordinator?](#expectations)
+
+
+<!-- /MarkdownTOC -->
+<a name="role-of-a-coordinator"></a>
+## Role of the Coordinator
+
+The Coordinator is not the owner of the component but much more the curator. 
```

Comment:
Why not change this name to curator then? :sweat_smile:

Reply:
Good point indeed this was in discussion, but we decided against, since we believe it hard to understand (curator might not be such a commonly known word). However, there have been voices prefering "curator". What would you prefer?

Reply:
We will ask the JF about the naming, your point is certainly valid.

Reply:
- Today contributors can prepare Pull Requests. How is this different but in the size of the Pull Request? If it is not different, there is no need for another model. If it is different this should be spelled out. 

- How could this fudge of responsibility with regard to bug fixing be beneficial? I am not convinced that hoping for high quality contributions can be considered a plan. Does a contribution warrant any responsibility and if so which and for how long? 

- How is the incentive issues dealt with i.e. how is ensured that coordinator do not end up with a lot of bug fixing and contributors get all the revenue? Do contributors have to get to the coordinator first and clear their plan or do they just work away and show up with code? 

- I thought the TB is planning for breaking up huge components. Why is this no longer a valid approach? It would keep the responsibility unfidged at least. 

- After reading and re-reading the text, I am still at a loss what the specific expected benefits are and in which specific cases this model should apply. 

- Can coordinators refuse PR that do not comply with their vision / plans for the component point blanc or do they need to substantiate and if so how?

Reply:
@atoedt:

We discussed your questions and remarks in the TB, these are our answers:

> Today contributors can prepare Pull Requests. How is this different but in the size of the Pull Request? If it is not different, there is no need for another model. If it is different this should be spelled out.

The crucial difference is, that the maintainer is a sovereign for his component. He decides what's possible and what's not possible. He may lay down rules, but is not obliged to do so. There are global guidelines he needs to respect, but besides these he has a lot of freedom and room. A contributor completely relies on the courtesy of the maintainer. This is not entirely negative and has advantages (e.g. responsiblity for bugs).

For some components in the software this model is problematic. This is especially the case when it comes to cross-section functionality that everyone has to life with. Coordinators have less freedom and are called upon to lay down rules for others and themselfs, so that contributors can develop an expectation about the component and participate reliably in the development. E.g. is it not possible for @Amstutz and @klees to reject some UI-component without a clear justification. The introduction of UI-components is governed by rules and processes and not the discretion of the two.

In that regard, the coordinator is much less a sovereign, king or owner of a component, but a facilitator for a cooperative development.

**TL/DR:** We define coordinators as a new role and a new maintenance model. Pull request are only an instrument to suggest code changes. We need the new maintenance model for parts of the code where the classic maintainer model does not work well. This need does not depend on the size of pull requests.

> How could this fudge of responsibility with regard to bug fixing be beneficial? I am not convinced that hoping for high quality contributions can be considered a plan. Does a contribution warrant any responsibility and if so which and for how long?

The coordinator gets to decide which code is merged along the rules outline here and in the contribution guideline. He may decide not to merge contributions from people that he deems unreliable or that he suspects won't take part in maintenance of the component. He may also deny to merge code that he deems unreliable or where the risk cannot be assessed. Code will also be contributed in public, i.e. everyone has the possibility to take part in building a reliable component. This is more than some hope for high quality contributions or fudge of responsibility. Even today there is no warrant that a bug gets fixed. However, this should be mentioned clearly in the document to avoid future questions.

Regarding the "maintainer is responsible" principle, the responsibility in the model proposed here is distributed less clearly. On the other hand, the code is supported and maintained by more people, increasing the bus-factor. Also, the maintainer-model can easily collapse, if to much responsibility is borne by a single person.

> How is the incentive issues dealt with i.e. how is ensured that coordinator do not end up with a lot of bug fixing and contributors get all the revenue? Do contributors have to get to the coordinator first and clear their plan or do they just work away and show up with code?

The document already accounts for that question:

* The motivation of the coordinator is mostly driven by the need of a  reliable component for a certain aspect. Further the coordinator is probably the most attractive contractor for clients aiming to change aspects of the component due to the very indepth know how and the listing as coordinator.
* The coordinator MAY ask for funding to perform the review.
* The coordinator moderates the discussion on finding a vision and on the development.

Also see the answer to your bugfixing question.

> I thought the TB is planning for breaking up huge components. Why is this no longer a valid approach? It would keep the responsibility unfidged at least.

There is more than one or the other. The document clearly outlines the scenarios for which we consider this model to be beneficial.

> After reading and re-reading the text, I am still at a loss what the specific expected benefits are and in which specific cases this model should apply.

It is hard to understand that, we tried hard to devise an approachable and comprehensible document and consider the "scenarios"-section to answer your question already. Maybe the problem is that we want to use this model for cross-section functionality. Could it be, that the nature of that functionality is hard to grasp for a non-developer? Maybe we should add examples in the "scenarios"-section (e.g. UI-framework) that further clarify that section.

> Can coordinators refuse PR that do not comply with their vision / plans for the component point blanc or do they need to substantiate and if so how?

The rules from our [contribution guideline](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-3/docs/documentation/contributing.md) apply here as well. Maybe we should explicitly link them and also rephrase them in the document. I.e.: Yes, contributions can be rejected. The instances to escalate issues are described in the document.

The coordinator is called upon to define additional criteria and processes for the component (e.g. UI-components). This is the gain of this model: PRs cannot be rejected arbitrarily. This allows other developers to build an expectation about the chances of their PR.